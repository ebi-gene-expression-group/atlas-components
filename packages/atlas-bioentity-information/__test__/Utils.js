const mockBioentityProperties = () => {
  return [ {
    type: `go`,
    name: `Gene Ontology`,
    values: [
      {
        text: `Link of relevance 1`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 1
      },
      {
        text: `a Link of relevance 2`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 2
      },
      {
        text: `Link of relevance 3`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 3
      },
      {
        text: `b Link of relevance 2`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 2
      },
      {
        text: `Link of relevance 3- another`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 3
      },
      {
        text: `RNA polymerase II core promoter proximal region sequence-specific DNA binding`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0000978`,
        relevance: 0
      }, {
        text: `nucleic acid binding`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0003676`,
        relevance: 0
      }, {
        text: `intracellular`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0005622`,
        relevance: 0
      }, {
        text: `nucleus`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0005634`,
        relevance: 0
      }, {
        text: `transcription, DNA-templated`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0006351`,
        relevance: 0
      }, {
        text: `regulation of transcription, DNA-templated`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0006355`,
        relevance: 0
      }, {
        text: `zinc ion binding`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0008270`,
        relevance: 0
      }, {
        text: `metal ion binding`,
        url: `http://amigo.geneontology.org/amigo/term/GO%3A0046872`,
        relevance: 0
      }]
    },
    {
      type: `synonym`,
      name: `Synonym`,
      values: [
        {
          text: `Sftp1`,
          url: ``,
          relevance: 0
        },
        {
          text: `Sftp-1`,
          url: ``,
          relevance: 0
        },
        {
          text: `SP-A`,
          url: ``,
          relevance: 0
        },
        {
          text: `surfactant pulmonary associated protein A1`,
          url: ``,
          relevance: 0
        },
        {
          text: `SFTPA1`,
          url: ``,
          relevance: 0
        }
      ]
    },
    {
      type: `ensgene`,
      name: `Ensembl gene`,
      values: [
        {
          text: `ENSMUSG00000021789`,
          url: `http://www.ensemblgenomes.org/id-gene/ENSMUSG00000021789`,
          relevance: 0
        }
      ]
    },
    {
      type: `enstranscript`,
      name: `Ensembl transcript`,
      values: [
        {
          text: `ENSMUST00000170719`,
          url: `http://www.ensemblgenomes.org/id/ENSMUST00000170719`,
          relevance: 0
        },
        {
          text: `ENSMUST00000022314`,
          url: `http://www.ensemblgenomes.org/id/ENSMUST00000022314`,
          relevance: 0
        }
      ]
    }
  ]
}

export { mockBioentityProperties }