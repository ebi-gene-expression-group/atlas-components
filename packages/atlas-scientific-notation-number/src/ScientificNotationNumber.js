import React from 'react'
import PropTypes from 'prop-types'

const removeTrailingZeroes = (str) => str.replace(/(\d)0+$/, `$1`)
const removeLeadingPlus = (str) => str.replace(/^\+/, ``)

const ScientificNotationNumber = ({value, accuracy, style}) => {
  if (value >= 0.1 && value < 100000 || value === 0) {
    return <span>{value}</span>
  }

  const scientificNotationString= value.toExponential(accuracy)
  const mantissaExponent = scientificNotationString.split(`e`)
  const mantissa = removeTrailingZeroes(mantissaExponent[0])
  const exponent = removeLeadingPlus(mantissaExponent[1])

  return(
    <span style={style}>
      {(+mantissa !== 1) ? `${mantissa} Ã— ` : ``}10
      <span style={{verticalAlign: `super`}}>{exponent}</span>
    </span>
  )
}

ScientificNotationNumber.propTypes = {
  value: PropTypes.number.isRequired,
  accuracy: PropTypes.number,
  style: PropTypes.object
}

ScientificNotationNumber.defaultProps = {
  accuracy: 4,
  style: {}
}

export default ScientificNotationNumber
