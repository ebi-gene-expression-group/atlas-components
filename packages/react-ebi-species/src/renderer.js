"use strict";

//*------------------------------------------------------------------*

var React = require('react');
var ReactDOM = require('react-dom');

//*------------------------------------------------------------------*

var Icon = require('./SpeciesIcon.jsx');

//*------------------------------------------------------------------*

module.exports = function(mountNode,species,colourOverride, colourPerKingdom) {
    ReactDOM.render(
        React.createElement(Icon,{
          species: species,
          colourOverride: colourOverride,
          colourPerKingdom: colourPerKingdom
        }), mountNode
    );
};
