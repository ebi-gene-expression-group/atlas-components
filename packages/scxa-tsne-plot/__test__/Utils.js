import HighchartsSeriesGenerator from 'highcharts-series-generator'

const seriesNames = [`0`, `1`, `2`, `3`, `4`]
const maxPointsPerSeries = 1000

const gradientColourRanges = () => {
  return [
    {
      colour: `rgb(215, 255, 255)`,
      threshold: 0,
      stopPosition: 0
    },
    {
      colour: `rgb(128, 255, 255)`,
      threshold: 10,
      stopPosition: 20
    },
    {
      colour: `rgb(0, 85, 225)`,
      threshold: 100,
      stopPosition: 40
    },
    {
      colour: `rgb(0, 0, 115)`,
      threshold: 10000,
      stopPosition: 100
    }
  ]
}

const plotData = (series) => {
  const allPoints = series.reduce((acc, series) => acc.concat(series.data), [])

  return {
    series: series,
    min: Math.min(...allPoints.map((point) => point.expressionLevel)),
    max: Math.max(...allPoints.map((point) => point.expressionLevel)),
    unit: `TPM`
  }
}

const randomHighchartsSeries = () => {
  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries)
}

const randomHighchartsSeriesWithNamesAndMaxPoints = (seriesNames, maxPointsPerSeries) => {
  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries)
}

const randomHighchartsSeriesWithSeed = () => {
  const seed = `A hair, Morty. I need one of your hairs. This isnâ€™t Game of Thrones.`

  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries, seed)
}

export { plotData, gradientColourRanges, randomHighchartsSeries, randomHighchartsSeriesWithSeed, randomHighchartsSeriesWithNamesAndMaxPoints}