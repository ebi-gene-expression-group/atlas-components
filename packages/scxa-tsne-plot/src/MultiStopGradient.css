.gradient {
    width: 20px;
    vertical-align: middle;
    margin: auto;
}

.tick {
    position: absolute;
    height: 2px;
    width: 20px;
}

.tickLabel {
    position: absolute;
    margin-top: -12px;
}

.left {
    composes: tickLabel;
    right: 20px;
    padding-right: 5px;
}

.right {
    composes: tickLabel;
    left: 20px;
    padding-left: 5px;
}