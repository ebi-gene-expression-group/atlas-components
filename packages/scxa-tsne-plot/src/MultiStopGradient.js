import React from 'react'
import PropTypes from 'prop-types'

import ScientificNotationNumber from 'expression-atlas-number-format'

import './util/MathRound'

// ranges have at least threshold (in ascending order) and stopPosition [0..100]
const putInRange = (ranges, val) => {
  if (val <= ranges[0].threshold) {
    return 0
  }

  if (val >= ranges[ranges.length - 1].threshold) {
    return 1
  }

  const rangeIndex = ranges.findIndex((range) => range.threshold >= val) - 1

  const withinRangeOffset =
    (val - ranges[rangeIndex].threshold) / (ranges[rangeIndex + 1].threshold - ranges[rangeIndex].threshold)

  return ranges[rangeIndex].stopPosition * (1 + withinRangeOffset) / 100
}



const MultiStopGradient = ({height, colourRanges, plotData}) => {
  const bg = colourRanges.map((colourRange) => `${colourRange.colour} ${colourRange.stopPosition}%`).join(`, `)
  const minMaxExpressionTickHeight = 2
  const gradientHeight = height - 100

  const minExpressionBottom =
    Math.min(
      gradientHeight - minMaxExpressionTickHeight,
      gradientHeight - gradientHeight * putInRange(colourRanges, plotData.min))
  const maxExpressionBottom =
    Math.max(
      minMaxExpressionTickHeight,
      gradientHeight - gradientHeight * putInRange(colourRanges, plotData.max))

  return (
    <div className={`small-2 columns text-center`}>
      <div>
        <small><ScientificNotationNumber value={colourRanges[colourRanges.length - 1].threshold}/> {plotData.unit}</small>
      </div>

      <div
        style={{
        width: `20px`,
        height: `${gradientHeight}px`,
        background: `linear-gradient(0deg, ${bg})`,
        verticalAlign: `middle`,
        margin: `auto`}}>

        <div style={{position: `relative`, height: `2px`, width: `20px`, background: `grey`, top: `${maxExpressionBottom}px`}}>
          <div style={{position: `absolute`, marginLeft: `24px`, marginTop: `-0.75rem`}}>
            <small><ScientificNotationNumber value={Math.round10(plotData.max, -2)}/></small>
          </div>
        </div>
        <div style={{position: `relative`, height: `2px`, width: `20px`, background: `grey`, top: `${minExpressionBottom}px`}}>
          <div style={{position: `absolute`, marginLeft: `24px`, marginTop: `-0.75rem`}}>
            <small><ScientificNotationNumber value={Math.round10(plotData.min, -2)}/></small>
          </div>
        </div>

      </div>

      <div>
        <small><ScientificNotationNumber value={colourRanges[0].threshold}/> {plotData.unit}</small>
      </div>
    </div>
  )
}

MultiStopGradient.propTypes = {
  height: PropTypes.number.isRequired,
  colourRanges: PropTypes.arrayOf(PropTypes.shape({
    colour: PropTypes.string.isRequired,
    threshold: PropTypes.number.isRequired
  })).isRequired,
  plotData: PropTypes.shape({
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired,
    unit: PropTypes.string.isRequired
  })
}

export default MultiStopGradient
