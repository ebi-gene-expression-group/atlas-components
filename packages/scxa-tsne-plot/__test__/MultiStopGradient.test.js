import React from 'react'
import renderer from 'react-test-renderer'

import Enzyme from 'enzyme'
import {shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import MultiStopGradient from '../src/MultiStopGradient'
import ScientificNotationNumber from 'expression-atlas-number-format'
import {gradientColourRanges, randomHighchartsSeries, randomHighchartsSeriesWithSeed, plotData} from './Utils'

Enzyme.configure({ adapter: new Adapter() })

describe(`MultiStopGradient`, () => {
  const GRADIENT_HEIGHT = 600

  test(`with random data matches snapshot`, () => {
    const randomSeries = randomHighchartsSeriesWithSeed()

    const tree = renderer
      .create(<MultiStopGradient height={GRADIENT_HEIGHT} showTicks={true} colourRanges={gradientColourRanges()} plotData={plotData(randomSeries)}/>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test(`displays 4 ticks, 2 on the right and 2 on the left`, () => {
    const randomSeries = randomHighchartsSeries()
    const wrapper = mount(<MultiStopGradient height={GRADIENT_HEIGHT} showTicks={true} colourRanges={gradientColourRanges()} plotData={plotData(randomSeries)}/>)

    expect(wrapper.find('.tick').length).toBe(4)
    expect(wrapper.find('.tick .right').length).toBe(2)
    expect( wrapper.find('.tick .left').length).toBe(2)
  })

  test(`min tick has a lower value and position than max tick`, () => {
    const randomSeries = randomHighchartsSeries()
    const data = plotData(randomSeries)
    const wrapper = mount(<MultiStopGradient height={GRADIENT_HEIGHT} showTicks={true} colourRanges={gradientColourRanges()} plotData={data}/>)

    // The max tick should be positioned higher than the min tick
    const maxTick = wrapper.find('.tick .left').parent().get(0)
    const minTick = wrapper.find('.tick .left').parent().get(0)

    expect(parseInt(maxTick.props.style.top)).toBeGreaterThanOrEqual(parseInt(minTick.props.style.top))

    // The first left tick should be the max expression value, the second the min expression value
    const maxTickNumber = wrapper.find('.tick .left').find(ScientificNotationNumber).get(0)
    const minTickNumber = wrapper.find('.tick .left').find(ScientificNotationNumber).get(1)

    expect(maxTickNumber.props.value).toBe(Math.round10(data.max, -2))
    expect(minTickNumber.props.value).toBe(Math.round10(data.min, -2))
  })

})
