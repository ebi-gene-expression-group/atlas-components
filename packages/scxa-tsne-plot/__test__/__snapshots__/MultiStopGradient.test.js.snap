// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiStopGradient with random data matches snapshot 1`] = `
<div
  className="small-2 columns text-center"
>
  <small>
    <span>
      10000
    </span>
     
    TPM
  </small>
  <div
    className="gradient"
    style={
      Object {
        "background": "linear-gradient(0deg, rgb(215, 255, 255) 0%, rgb(128, 255, 255) 20%, rgb(0, 85, 225) 40%, rgb(0, 0, 115) 100%)",
        "height": "500px",
      }
    }
  >
    <div
      className="dimgray"
      style={
        Object {
          "top": "26px",
        }
      }
    >
      <small
        className="left"
      >
        <span>
          9997.45
        </span>
      </small>
    </div>
    <div
      className="dimgray"
      style={
        Object {
          "top": "487.57040149782796px",
        }
      }
    >
      <small
        className="left"
      >
        <span>
          3.64
        </span>
      </small>
    </div>
    <div
      className="lightgray"
      style={
        Object {
          "top": "424px",
        }
      }
    >
      <small
        className="right"
      >
        <span>
          10
        </span>
      </small>
    </div>
    <div
      className="lightgray"
      style={
        Object {
          "top": "324px",
        }
      }
    >
      <small
        className="right"
      >
        <span>
          100
        </span>
      </small>
    </div>
  </div>
  <div>
    <small>
      <span>
        0
      </span>
       
      TPM
    </small>
  </div>
</div>
`;
